File: TMG.EMME/TMGToolbox2/src/Assign/assign_traffic.py
# ---LICENSE----------------------
"""
    Copyright 2022 Travel Modelling Group, Department of Civil Engineering, University of Toronto

    This file is part of the TMG Toolbox.

    The TMG Toolbox is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    The TMG Toolbox is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with the TMG Toolbox.  If not, see <http://www.gnu.org/licenses/>.
"""

# ---METADATA---------------------
"""
Toll-Based Road Assignment

    Authors: David King, Eric Miller

    Latest revision by: dKingII
    
    Executes a multi-class road assignment which allows for the generalized penalty of road tolls.
    
    V 1.0.0

    V 1.1.0 Added link volume attributes for increased resolution of analysis.

    V 1.1.1 Updated to allow for multi-threaded matrix calcs in 4.2.1+

    V 2.0.0 Refactored to work with XTMF2/TMGToolbox2 on 2021-10-21 by williamsDiogu      

    V 2.0.1 Updated to receive JSON object parameters from XTMX2

    V 2.0.2 Updated to receive JSON file parameters from Python API call

"""
import inro.modeller as _m
import multiprocessing

_m.InstanceType = object
_m.ListType = list
_m.TupleType = object

_trace = _m.logbook_trace
_MODELLER = _m.Modeller()  # Instantiate Modeller once.
_util = _MODELLER.module("tmg2.utilities.general_utilities")
EMME_VERSION = _util.get_emme_version(tuple)

matrix_calc_tool = _MODELLER.tool("inro.emme.matrix_calculation.matrix_calculator")
network_calculation_tool = _MODELLER.tool("inro.emme.network_calculation.network_calculator")
traffic_assignment_tool = _MODELLER.tool("inro.emme.traffic_assignment.sola_traffic_assignment")
extra_parameter_tool = _MODELLER.tool("inro.emme.traffic_assignment.set_extra_function_parameters")
delete_matrix = _MODELLER.tool("inro.emme.data.matrix.delete_matrix")


class AssignTraffic(_m.Tool()):
    version = "2.0.2"
    tool_run_msg = ""
    # For progress reporting, enter the integer number of tasks here
    number_of_tasks = 4
    # Tool Input Parameters
    #    Only those parameters necessary for Modeller and/or XTMF to dock with
    #    need to be placed here. Internal parameters (such as lists and dicts)
    #    get initialized during construction (__init__)
    # Parameters can takes in a json file name depending on entry point(either through XTMF or api calls )
    parameters = _m.Attribute(str)
    number_of_processors = _m.Attribute(int)

    def __init__(self):
        self._tracker = _util.progress_tracker(self.number_of_tasks)
        self.scenario = _MODELLER.scenario
        self.number_of_processors = multiprocessing.cpu_count()
        self._traffic_util = _util.assign_traffic_util()

    def page(self):
        pb = _m.ToolPageBuilder(
            self,
            title="Multi-Class Road Assignment",
            description="Cannot be called from Modeller.",
            runnable=False,
            branding_text="XTMF",
        )
        return pb.render()

    def __call__(self, parameters):
        scenario = _util.load_scenario(parameters["scenario_number"])
        try:
            self._execute(scenario, parameters)
        except Exception as e:
            raise Exception(_util.format_reverse_stack())

    def run_xtmf(self, parameters):
        scenario = _util.load_scenario(parameters["scenario_number"])

        try:
            self._execute(scenario, parameters)
        except Exception as e:
            raise Exception(_util.format_reverse_stack())

    def _execute(self, scenario, parameters):
        """
        Definition of common names:
            - matrix_name (type(str)): e.g. cost_matrix, demand_matrix etc.
            - matrix_id (type(str)): e.g. mf0, mf2 etc.
                Note - matrix id expects either of ms, mo, md, or mf, before the number.
            - temp_matrix_list: keeps track (by appending to it) of all temporary matrices
                created and deletes them with the help of a temp_matrix_manager() at the at
                the end of each run (including when code catches an error)
            - temp_attributes_list: keeps track of all temporary attributes created and deletes
                them at the at the end of each run (including when code catches an error)
        """
        load_input_matrix_list = self._traffic_util.load_input_matrices(parameters, "demand_matrix")
        load_output_matrix_dict = self._traffic_util.load_output_matrices(
            parameters,
            matrix_name=["cost_matrix", "time_matrix", "toll_matrix"],
        )
        with _trace(
            name="%s (%s v%s)" % (parameters["run_title"], self.__class__.__name__, self.version),
            attributes=self._traffic_util.load_atts(scenario, parameters, self.__MODELLER_NAMESPACE__),
        ):
            self._tracker.reset()
            with _util.temporary_matrix_manager() as temp_matrix_list:
                demand_matrix_list = self._traffic_util.init_input_matrices(load_input_matrix_list, temp_matrix_list)
                cost_matrix_list = self._traffic_util.init_output_matrices(
                    load_output_matrix_dict,
                    temp_matrix_list,
                    matrix_name="cost_matrix",
                    description="",
                )
                time_matrix_list = self._traffic_util.init_output_matrices(
                    load_output_matrix_dict,
                    temp_matrix_list,
                    matrix_name="time_matrix",
                    description="",
                )
                toll_matrix_list = self._traffic_util.init_output_matrices(
                    load_output_matrix_dict,
                    temp_matrix_list,
                    matrix_name="toll_matrix",
                    description="",
                )
                peak_hour_matrix_list = self._traffic_util.init_temp_peak_hour_matrix(parameters, temp_matrix_list)
                self._tracker.complete_subtask()

                with _util.temporary_attribute_manager(scenario) as temp_attribute_list:
                    time_attribute_list = self._traffic_util.create_time_attribute_list(
                        scenario, demand_matrix_list, temp_attribute_list
                    )
                    cost_attribute_list = self._traffic_util.create_cost_attribute_list(
                        scenario, demand_matrix_list, temp_attribute_list
                    )
                    transit_attribute_list = self._traffic_util.create_transit_traffic_attribute_list(
                        scenario, demand_matrix_list, temp_attribute_list
                    )
                    # Create volume attributes
                    for tc in parameters["traffic_classes"]:
                        self._traffic_util.create_volume_attribute(scenario, tc["volume_attribute"])
                    # Calculate transit background traffic
                    self._traffic_util.calculate_transit_background_traffic(
                        scenario,
                        parameters,
                        self._tracker,
                    )
                    # Calculate applied toll factor
                    applied_toll_factor_list = self._traffic_util.calculate_applied_toll_factor(parameters)
                    # Calculate link costs
                    self._traffic_util.calculate_link_cost(
                        scenario,
                        parameters,
                        demand_matrix_list,
                        applied_toll_factor_list,
                        cost_attribute_list,
                        self._tracker,
                    )
                    # Calculate peak hour matrix
                    self._traffic_util.calculate_peak_hour_matrices(
                        scenario,
                        parameters,
                        demand_matrix_list,
                        peak_hour_matrix_list,
                        self._tracker,
                        self.number_of_processors,
                    )
                    self._tracker.complete_subtask()

                    # Assign traffic to road network
                    with _m.logbook_trace("Running Road Assignments."):
                        completed_path_analysis = False
                        if completed_path_analysis is False:
                            attributes = self._traffic_util.load_attribute_list(parameters, demand_matrix_list)
                            attribute_list = attributes[0]
                            volume_attribute_list = attributes[1]
                            mode_list = self._traffic_util.load_mode_list(parameters)

                            sola_spec = self._traffic_util.get_primary_SOLA_spec(
                                demand_matrix_list,
                                peak_hour_matrix_list,
                                applied_toll_factor_list,
                                mode_list,
                                volume_attribute_list,
                                cost_attribute_list,
                                time_matrix_list,
                                attribute_list,
                                None,
                                None,
                                None,
                                None,
                                None,
                                None,
                                None,
                                parameters,
                                multiprocessing,
                            )
                            report = self._tracker.run_tool(traffic_assignment_tool, sola_spec, scenario=scenario)
                        checked = self._load_stopping_criteria(report)
                        number = checked[0]
                        stopping_criterion = checked[1]
                        value = checked[2]

                        print("Primary assignment complete at %s iterations." % number)
                        print("Stopping criterion was %s with a value of %s." % (stopping_criterion, value))

    def _load_stopping_criteria(self, report):
        stopping_criterion = report["stopping_criterion"]
        iterations = report["iterations"]
        if len(iterations) > 0:
            final_iteration = iterations[-1]
        else:
            final_iteration = {"number": 0}
            stopping_criterion == "MAX_ITERATIONS"
        number = final_iteration["number"]
        if stopping_criterion == "MAX_ITERATIONS":
            value = final_iteration["number"]
        elif stopping_criterion == "RELATIVE_GAP":
            value = final_iteration["gaps"]["relative"]
        elif stopping_criterion == "NORMALIZED_GAP":
            value = final_iteration["gaps"]["normalized"]
        elif stopping_criterion == "BEST_RELATIVE_GAP":
            value = final_iteration["gaps"]["best_relative"]
        else:
            value = "undefined"
        return number, stopping_criterion, value

    @_m.method(return_type=_m.TupleType)
    def percent_completed(self):
        return self._tracker.get_progress()

    @_m.method(return_type=str)
    def tool_run_msg_status(self):
        return self.tool_run_msg

File: TMG.EMME/TMG.EMME/Assign/AssignTraffic.cs
/*
    Copyright 2021 University of Toronto

    This file is part of TMG.EMME for XTMF2.

    TMG.EMME for XTMF2 is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    TMG.EMME for XTMF2 is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with TMG.EMME for XTMF2.  If not, see <http://www.gnu.org/licenses/>.
*/

using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using TMG.Emme;
using XTMF2;

namespace TMG.Emme.Assign
{
    [Module(Name = "Assign Demand To Road Network", Description = "",
        DocumentationLink = "http://tmg.utoronto.ca/doc/2.0")]
    public class AssignTraffic : BaseAction<ModellerController>
    {
        [Parameter(Name = "Background Transit", DefaultValue = "true", Description = "Set this to false to not assign transit vehicles on the roads",
            Index = 0)]
        public IFunction<string> BackgroundTransit;

        [Parameter(Name = "br Gap", DefaultValue = "0", Description = "The minimum gap required to terminate the algorithm.",
            Index = 1)]
        public IFunction<float> brGap;

        [Parameter(Name = "Iterations", DefaultValue = "100", Description = "The maximum number of iterations to run.",
            Index = 2)]
        public IFunction<int> Iterations;

        [Parameter(Name = "norm Gap", DefaultValue = "0", Description = "The minimum gap required to terminate the algorithm.",
            Index = 3)]
        public IFunction<float> normGap;

        [Parameter(Name = "Performance Flag", DefaultValue = "true", Description = "Set this to false to leave a free core for other work",
            Index = 5)]
        public IFunction<bool> PerformanceFlag;

        [Parameter(Name = "r Gap", DefaultValue = "0", Description = "The minimum gap required to terminate the algorithm.",
            Index = 6)]
        public IFunction<float> rGap;

        [Parameter(Name = "Run Title", DefaultValue = "road assignment", Description = "The name of the run to appear in the logbook.",
            Index = 7)]
        public IFunction<string> RunTitle;

        [Parameter(Name = "Scenario Number", DefaultValue = "0", Description = "The scenario number to execute against.",
            Index = 8)]
        public IFunction<int> ScenarioNumber;

        [Parameter(Name = "SOLA Flag", DefaultValue = "true", Description = "Sola flag",
            Index = 9)]
        public IFunction<bool> SOLAFlag;

        [SubModule(Name = "Traffic Classes", Description = "", Index = 10)]
        public IFunction<TrafficClass>[] TrafficClasses;

        [Module(Name = "Traffic Class", Description = "",
        DocumentationLink = "http://tmg.utoronto.ca/doc/2.0")]
        public class TrafficClass : XTMF2.IModule
        {
            [Parameter(Name = "Mode", DefaultValue = "c", Description = "The mode for this class.",
                Index = 0)]
            public IFunction<char> Mode;

            [Parameter(Name = "Demand Matrix", DefaultValue = "0", Description = "The id of the demand matrix to use.",
                Index = 1)]
            public IFunction<string> DemandMatrixNumber;

            [Parameter(Name = "Time Matrix", DefaultValue = "0", Description = "The matrix number to save in vehicle travel times",
                Index = 2)]
            public IFunction<string> TimeMatrix;

            [Parameter(Name = "Cost Matrix", DefaultValue = "0", Description = "The matrix number to save the total cost into.",
                Index = 3)]
            public IFunction<string> CostMatrix;

            [Parameter(Name = "Toll Matrix", DefaultValue = "0", Description = "The matrix to save the toll costs into.",
                Index = 4)]
            public IFunction<string> TollMatrix;

            [Parameter(Name = "Peak Hour Factor", DefaultValue = "1", Description = "A factor to apply to the demand in order to build a representative hour.",
            Index = 4)]
            public IFunction<float> PeakHourFactor;

            [Parameter(Name = "Volume Attribute", DefaultValue = " @auto_volume1", Description = "The name of the attribute to save the volumes into (or None for no saving).",
                Index = 5)]
            public IFunction<string> VolumeAttribute;

            [Parameter(Name = "Toll Attribute ID", DefaultValue = " @toll", Description = "The attribute containing the road tolls for this class of vehicle.",
                Index = 6)]
            public IFunction<string> LinkTollAttributeID;

            [Parameter(Name = "Toll Weight", DefaultValue = "0", Description = "The toll weight",
                Index = 7)]
            public IFunction<float> TollWeight;

            [Parameter(Name = "Link Cost", DefaultValue = "0", Description = "The penalty in minutes per dollar to apply when traversing a link.",
                Index = 8)]
            public IFunction<float> LinkCost;

            [SubModule(Name = "Path Analysis", Description = "", Index = 9, Required = false)]
            public IFunction<PathAnalysis>[] PathAnalyses;

            public string Name { get; set; }

            public bool RuntimeValidation(ref string error)
            {
                return true;
            }

            public void WriteParameters(System.Text.Json.Utf8JsonWriter writer)
            {
                writer.WriteStartObject();
                writer.WriteString("name", Name);
                writer.WriteString("mode", Mode.Invoke().ToString());
                writer.WriteString("demand_matrix", DemandMatrixNumber.Invoke());
                writer.WriteString("time_matrix", TimeMatrix.Invoke());
                writer.WriteString("cost_matrix", CostMatrix.Invoke());
                writer.WriteString("toll_matrix", TollMatrix.Invoke());
                writer.WriteNumber("peak_hour_factor", PeakHourFactor.Invoke());
                writer.WriteString("volume_attribute", VolumeAttribute.Invoke());
                writer.WriteString("link_toll_attribute_id", LinkTollAttributeID.Invoke());
                writer.WriteNumber("toll_weight", TollWeight.Invoke());
                writer.WriteNumber("link_cost", LinkCost.Invoke());
                writer.WriteStartArray("path_analyses");
                foreach (var pathAnalysis in PathAnalyses)
                {
                    pathAnalysis.Invoke().WriteParameters(writer);
                }
                writer.WriteEndArray();
                writer.WriteEndObject();
            }
        }

        [Module(Name = "Path Analysis", Description = "",
        DocumentationLink = "http://tmg.utoronto.ca/doc/2.0")]
        public class PathAnalysis : XTMF2.IModule
        {
            [Parameter(Name = "Attribute ID", DefaultValue = "0", Description = "The attribute to use for analysis",
                Index = 0)]
            public IFunction<string> AttributeId;

            [Parameter(Name = "Aggregation Matrix", DefaultValue = "0", Description = "The matrix number to store the results into",
                Index = 1)]
            public IFunction<int> AggregationMatrix;

            [Parameter(Name = "Operator", DefaultValue = "+", Description = "The operator to use to aggregate the matrix. Example:'+' for emissions",
                Index = 2)]
            public IFunction<string> AggregationOperator;

            [Parameter(Name = "Lower Bound for Path Selector", DefaultValue = "None", Description = "Lower Bound for Path Selector None The number to use for the lower bound in path selection",
                Index = 3)]
            public IFunction<string> LowerBound;

            [Parameter(Name = "Upper Bound for Path Selector", DefaultValue = "None", Description = "Upper Bound for Path Selector None The number to use for the upper bound in path selection",
                Index = 4)]
            public IFunction<string> UpperBound;

            [Parameter(Name = "Path Selection", DefaultValue = "ALL", Description = "The paths that will be used for analysis",
                Index = 5)]
            public IFunction<string> PathSelection;

            [Parameter(Name = "Multiply Path Prop By Demand", DefaultValue = "true", Description = "Choose whether to multiply the path proportion by the analyzed demand",
                Index = 6)]
            public IFunction<string> MultiplyPathPropByDemand;

            [Parameter(Name = "Multiply Path Prop By Value", DefaultValue = "true", Description = "Choose whether to multiply the path proportion by the path value",
                Index = 7)]
            public IFunction<string> MultiplyPathPropByValue;

            [Parameter(Name = "Analysis Attributes", DefaultValue = "@auto_volume", Description = "The analysis attributes",
                Index = 8)]
            public IFunction<string> AnalysisAttributes;

            [Parameter(Name = "Analysis Attributes Matrix Id", DefaultValue = "0", Description = "The matrix number to store the results into",
                Index = 21)]
            public IFunction<string> AnalysisAttributesMatrixId;

            public string Name { get; set; }

            public bool RuntimeValidation(ref string error)
            {
                return true;
            }

            public void WriteParameters(System.Text.Json.Utf8JsonWriter writer)
            {
                writer.WriteStartObject();
                writer.WriteString("attribute_id", AttributeId.Invoke());
                writer.WriteNumber("aggregation_matrix", AggregationMatrix.Invoke());
                writer.WriteString("aggregation_operator", AggregationOperator.Invoke());
                writer.WriteString("lower_bound", LowerBound.Invoke());
                writer.WriteString("upper_bound", UpperBound.Invoke());
                writer.WriteString("path_selection", PathSelection.Invoke());
                writer.WriteString("multiply_path_prop_by_demand", MultiplyPathPropByDemand.Invoke());
                writer.WriteString("multiply_path_prop_by_value", MultiplyPathPropByValue.Invoke());
                writer.WriteString("analysis_attributes", AnalysisAttributes.Invoke());
                writer.WriteString("analysis_attributes_matrix", AnalysisAttributesMatrixId.Invoke());
                writer.WriteEndObject();
            }
        }

        public override void Invoke(ModellerController context)
        {
            context.Run(this, "tmg2.Assign.assign_traffic", JSONParameterBuilder.BuildParameters(writer =>
            {
                writer.WriteNumber("scenario_number", ScenarioNumber.Invoke());
                writer.WriteNumber("iterations", Iterations.Invoke());
                writer.WriteNumber("r_gap", rGap.Invoke());
                writer.WriteNumber("br_gap", brGap.Invoke());
                writer.WriteNumber("norm_gap", normGap.Invoke());
                writer.WriteBoolean("performance_flag", PerformanceFlag.Invoke());
                writer.WriteBoolean("sola_flag", SOLAFlag.Invoke());
                writer.WriteString("background_transit", BackgroundTransit.Invoke());
                writer.WriteString("run_title", RunTitle.Invoke());
                writer.WriteStartArray("traffic_classes");
                foreach (var trafficClass in TrafficClasses)
                {
                    trafficClass.Invoke().WriteParameters(writer);
                }
                writer.WriteEndArray();
            }), LogbookLevel.Standard);
        }

    }
}
